### PR TITLE
Refactor the authentication flow in the `authenticate` function

### PR DESCRIPTION
**Pull request by CFIALeronB**
- Transitioned from a state-based logic (INITIAL_STATE, AWAITING_EMAIL, EMAIL_SENT, USER_AUTHENTICATED) to a request-type-based logic.
- The new approach focuses on the type of request, specifically the presence and type of token in the request.
- Removed state transition functions previously used.
- Added test cases for JWT validation.

#27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,22 +1,54 @@
...
-4. **Seamless Redirection**: After verification, you'll be automatically redirected to the original client application, now authenticated.
-
+Louis Login Backend is a centralized authentication system designed for Single Sign-On (SSO) functionality. It streamlines user authentication across multiple client applications by providing a seamless flow for token-based email verification.
+
+## Workflow Overview:
+
+### 1. Client Application Check
+* The user accesses the client application website.
+* The application checks for a session cookie to determine if the user is already authenticated.
+  * If a session cookie exists: The user is redirected to the client application's main page.
```
Comment (rngadam) [inline README.md:10]: * check if lifetime is still valid

File: tests/test_authentication_flow.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+"""
+Verify Token Test Cases.
+"""
+from datetime import timedelta, datetime
+from flask.testing import FlaskClient
+import pytest
```
Comment (rngadam) [inline tests/test_authentication_flow.py:6]: future issues:

https://github.com/ai-cfia/github-workflows/blob/main/.github/workflows/workflow-lint-test-python.yml

should we revisit unittest vs pytest?

install and run ruff (VSC ruff extension)

**Review by rngadam — APPROVED** [review]
please do rename the tests before merging and create the future work issues

File: tests/test_authentication_flow.py (hunk 2 above)
Comment (CFIALeronB) [reply tests/test_authentication_flow.py:6]: #30

File: tests/conftest.py [hunk 3]
```diff
@@ -1,13 +1,17 @@
...
+from pathlib import Path
 import pytest
 from flask import Flask
+from jwt_utils import generate_email_verification_token
 from generate_jwt import generate_jwt
 
 # Import the Flask application instance from your app module
 from app import app as flask_app
 
+SERVER_PRIVATE_KEY = Path('tests/server_private_key/server_private_key.pem')
```
[the remark this answers is not in the document]
Comment (CFIALeronB) [reply tests/conftest.py:13]: #25

**Pull request merged**